Step1 문자열 덧셈 계산기 PR

louisJu opened the pull request:
안녕하세요 리뷰어님!  
1단계 문자열 덧셈 계산기 PR 요청드립니다.

이번에도 잘 부탁드립니다 :) 
새해복 많이 받으세요~

Commit:
fix(racingcar): step0 코멘트 내용 적용

1. 사용하지 않는 코드(주석처리) 제거
2. exception 메시지 추가
3. functionalInterface annotation을 이용한 명시적 선언

Commit:
feat(stringcaculator): 빈문자열 또는 null을 입력한 경우 0을 반환

Commit:
feat(stringcaculator): 요구사항에 대한 개발
1. 기본 구분자로 문자열 처리
2. 커스텀 구분자로 문자열 처리
3. 숫자이외의 값 또는 음수 전달시 RuntimeException 예외 throw

dkswnkk commented on src/main/java/stringcaculator/Calculator.kt:
```diff
@@ -0,0 +1,38 @@
+        val matcher = Pattern.compile(CUSTOM_DELIMITER_REGEX).matcher(text)
+        val nums = when {
+            matcher.find() -> {
+                val customDelimiter = matcher.group(1)
+                split(matcher.group(2), customDelimiter)
```
group(1)과 group(2)의 매직넘버들도 상수화해볼 수 있을 것 같아요!

dkswnkk commented on src/main/java/stringcaculator/Number.kt:
```diff
@@ -0,0 +1,24 @@
+
+import java.lang.NumberFormatException
+
+data class Number(val num: String) {
```
Number 객체 생성👍

dkswnkk commented on src/main/java/stringcaculator/Number.kt:
```diff
@@ -0,0 +1,24 @@
+data class Number(val num: String) {
+
+    init {
+        try {
```
현재 예외 처리가 다소 복잡한 것 같아요. 아래와 같이 단순화해 볼 수도 있을 것 같아요!

## suggestion1
```suggestion
init {
    try {
        val intNum = num.toInt()
        checkNegative(intNum)
    } catch (e: Exception) {
        when (e) {
            is NumberFormatException -> throw RuntimeException("숫자타입만 입력가능합니다. $num")
            is RuntimeException -> throw RuntimeException("음수는 포함될 수 없습니다. $num")
            else -> throw e
        }
    }
}
```
## suggestion2
```suggestion
init {
    val intNum = num.toIntOrNull() ?: throw RuntimeException("숫자타입만 입력가능합니다. $num")
    if (intNum < 0) throw RuntimeException("음수는 포함될 수 없습니다. $num")
}

```

louisJu replied:
저도 이부분에서 비슷한 느낌을 받았는데... ㅎㅎ 의견 감사합니다!

dkswnkk commented on src/main/java/stringcaculator/Calculator.kt:
```diff
@@ -0,0 +1,38 @@
+        if (text.isNullOrEmpty()) {
+            return ZERO
+        }
+        val matcher = Pattern.compile(CUSTOM_DELIMITER_REGEX).matcher(text)
```
Pattern 객체들은 내부적으로 매번 새로운 인스턴스를 생성하기 때문에 상대적으로 생성 비용이 비싼 객체들입니다. 한번 개선해 보시면 좋을 것 같아요!

cc. https://zhenxikim.tistory.com/79

dkswnkk commented on src/main/java/stringcaculator/Calculator.kt:
```diff
@@ -0,0 +1,38 @@
+    val CUSTOM_DELIMITER_REGEX = "//(.)\n(.*)"
+
+
+    fun add(text: String?): Int {
```
현재 Calculator의 add() 함수가 너무 많은 역할을 담당하고 있는 것 같아요.
책임을 나누도록 개선해 볼 수 있을 것 같습니다!

dkswnkk commented on src/test/java/stringcaculator/CalculatorTest.kt:
```diff
@@ -0,0 +1,61 @@
+    @DisplayName("숫자 하나를 입력한 경우 해당 숫자를 반환한다.")
+    @Test
+    fun 숫자하나입력() {
+        val result = calculator.add("7")
```
@ParameterizedTest를 사용해 보시는 건 어떨까요!?

dkswnkk commented on src/main/java/stringcaculator/Calculator.kt:
```diff
@@ -0,0 +1,38 @@
+        return result
+    }
+
+    private fun split(text: String, vararg delimiters: String): List<Number> {
```
`List<Number>`를 일급 컬렉션으로 만들어 활용해보는 방법도 좋을 것 같아요!

dkswnkk reviewed (changes requested):
연휴에도 고생많으십니다. 👍
몇 가지 코멘트를 남겼으니 확인 부탁드려요!

Commit:
refactor(stringcaculator): 피드백에 대한 처리

1. Calculator의 add() 함수 책임 분리 (Calculator.kt)
2. Numbers 라는 일급 컬렉션 사용 (Numbers.kt)
3. 비용이 많이 드는 Pattern 상수화 (Calculator.kt)
4. 숫자 예외처리에 대한 복잡도 감소 (Number.kt)
5. macher.group()의 숫자 상수화 (Calculator.kt)
6. 테스트에 @ParameterizedTest 사용해보기 (CalculatorTest.kt)

louisJu commented:
안녕하세요! 
말씀주신 피드백 내용들 바탕으로 코드를 리팩토링 해보았습니다!
한번 더 피드백 부탁드립니다~ :)

dkswnkk commented on src/main/java/stringcaculator/Calculator.kt:
```diff
@@ -0,0 +1,28 @@
+        }
+
+        val (delimiter, body) = parseInput(text)
+        return Numbers.generateNumbers(body, delimiter).sum()
```
<img width="473" alt="image" src="https://github.com/user-attachments/assets/bb0b3084-2a0c-4974-a343-129610484977" />

Numbers에 대한 커밋이 누락된 것 같은데 한번 확인 부탁드려요! 컴파일 에러가 바생합니다!

louisJu replied:
😅 헛...! 이런 실수를... ㅠㅠ

dkswnkk commented on src/test/java/stringcaculator/CalculatorTest.kt:
```diff
@@ -0,0 +1,59 @@
+
+    @DisplayName("문자열에 음수가 전달되면 RuntimeException을 throw 한다.")
+    @Test
+    fun 문자열에_숫자이외의값_전달시_RuntimeException_throw() {
```
@DisplayName이 잘못된 것 같아요!

dkswnkk commented on src/test/java/stringcaculator/CalculatorTest.kt:
```diff
@@ -0,0 +1,59 @@
+
+    @DisplayName("숫자 사이에 컴마 또는 콜론 구분자 입력할 경우 숫자의 합을 반환한다.")
+    @Test
+    fun 문자열을_컴마_또는_콜론으로_구분하여_합을_반환() {
+        val result = calculator.add("1,2:3,4")
+        assertThat(result).isEqualTo(10)
+    }
```
검증부는 하드코딩을 해보는 것은 어떨까요? `@CsvSource`에 대해서도 한번 사용해 보시면 좋을 것 같아요😃

cc. https://jojoldu.tistory.com/615

dkswnkk commented on src/main/java/stringcaculator/Calculator.kt:
```diff
@@ -0,0 +1,28 @@
+        return Numbers.generateNumbers(body, delimiter).sum()
+    }
+
+    private fun parseInput(text: String): Pair<Array<String>, String> {
```
parse 부분도 사실 또 하나의 객체로 만들어 볼 수 있을 것 같아요. 그러면 어떤 장점이 있을까요?

앞으로 미션 진행하시면서 Wrapper Class 혹은 일급객체 사용하시면서 각각의 장단점을 직접 한번 느껴보시면 좋을 것 같아요.😄

louisJu replied:
함수 기능에 대한 책임 분리가 명확해 지고 코드의 응집도 가 높아지는 장점이 있을 것 같네요 :)  
이후 추가 요구사항에 대한 확장성 측면에서도 용이할 것 같아요~

dkswnkk reviewed (changes requested):
피드백 잘 반영해주셨습니다!

Numbers에 대한 커밋이 누락된 것 같은데 한번 확인 부탁드려요!

Commit:
feat(stringcaculator): 누락된 클래스 추가 - Numbers 라는 일급 컬렉션 사용 (Numbers.kt)

Commit:
fix(stringcaculator): 피드백 반영 2
1. DisplayName 오류 수정, @CsvSource 활용 (CalculatorTest.kt)
2. parse 처리 일급 객체 활용하여 책임 분리 (Parser.kt)

dkswnkk reviewed (approved):
어제가 설날이라 리뷰가 조금 늦었습니다🥲
이번 단계에서 값 객체를 만들어 책임을 부여했던 방법들은 이후에도 계속해서 사용될 개념이니 계속해서 이어갔으면 좋겠습니다😀

다음 단계 진행해 주세요!
